### PR TITLE
US66693: Refactor policy evaluation decision caching to allow eviction of keys

### PR DESCRIPTION
**Pull request by FrankGasparovic**
Signed-off-by: Anubhav <anubhav.saggi@ge.com>

### CODE REVIEW

**Comment by FrankGasparovic** [comment]
Notes on new caching: 
https://devcloud.swcoe.ge.com/devspace/display/RMOCM/ACS+Decision+Caching

File: service/src/main/java/com/ge/predix/acs/service/policy/evaluation/PolicyEvaluationServiceImpl.java [hunk 1]
```diff
@@ -150,7 +156,12 @@ public PolicyEvaluationResult evalPolicy(final PolicyEvaluationRequestV1 request
                             "Processed Policy Evaluation for: "
                                     + "resourceUri='%s', subjectIdentifier='%s', action='%s'," + " result='%s'",
                             uri, subjectIdentifier, action, result.getEffect()));
-            this.cache.set(key, result);
+            try {
+                this.cache.set(key, result);
+            } catch (Exception e) {
+                LOGGER.error(String.format(
+                        "Unable to set cache key '%s' to value '%s' due to redis exception", key, result), e);
+            }
```
Comment (irinaepshteyn) [inline service/src/main/java/com/ge/predix/acs/service/policy/evaluation/PolicyEvaluationServiceImpl.java]: Error message should not be tied to redis.  Could have other cache implementations.

File: service/src/main/java/com/ge/predix/acs/service/policy/evaluation/PolicyEvaluationServiceImpl.java [hunk 2]
```diff
@@ -118,7 +118,13 @@ public PolicyEvaluationResult evalPolicy(final PolicyEvaluationRequestV1 request
...
         }
 
-        PolicyEvaluationResult result = this.cache.get(key);
+        PolicyEvaluationResult result = null;
+        try {
+            result = this.cache.get(key);
+        } catch (Exception e) {
+            LOGGER.error(String.format("Unable to get cache key '%s'", key), e);
+        }
+
```
Comment (irinaepshteyn) [inline service/src/main/java/com/ge/predix/acs/service/policy/evaluation/PolicyEvaluationServiceImpl.java:124]: Add a unit test that policy evaluation returns expected result if cache get or set fails

File: service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java [hunk 3]
```diff
@@ -67,6 +88,11 @@ public PolicyEvaluationResult get(final PolicyEvaluationRequestCacheKey key) {
         } catch (IOException e) {
             throw new IllegalStateException("Failed to read policy evaluation result as JSON.", e);
         }
+        values.remove(values.size() - 1);
+
+        String zone = key.getZoneId();
+        values.addAll(multiGet(cachedResult.getResolvedResourceUris().stream()
+                .map(resourceUri -> resourceKey(zone, resourceUri)).collect(Collectors.toList())));
```
Comment (irinaepshteyn) [inline service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java]: Have you thought of including resolved URIs as part of evaluation result key to avoid another trip to the cache store?  Something like:
zoneId:resourceId:resolvedURIList:...
Are we limited on the length of the key?

File: service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java [hunk 4]
```diff
@@ -79,10 +105,12 @@ public PolicyEvaluationResult get(final PolicyEvaluationRequestCacheKey key) {
     }
 
     private List<String> assembleKeys(final PolicyEvaluationRequestCacheKey key) {
+        // Only need to assemble keys for policy sets, the subject, and the policy evaluation result.
+        // Resource-related information will be captured in the resolved resource URIs from the cached
+        // policy evaluation result.
         List<String> keys = new ArrayList<>();
```
Comment (irinaepshteyn) [inline service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java]: If resolvedURI were part of evaluation result key we could assemble resource keys for resolved URIs here.

File: service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java (hunk 3 above)
Comment (irinaepshteyn) [reply service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java]: Add a unit test that verifies that get returns null if one of the resolved resources change.  Resolved URI of the changed does not match resource URI in the request.

File: service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java [hunk 5]
```diff
@@ -121,22 +148,21 @@ public void set(final PolicyEvaluationRequestCacheKey key, final PolicyEvaluatio
...
 
-    @Override
-    public void setResourceTranslation(final String zoneId, final String fromResourceId, final String toResourceId) {
-        String fromKey = resourceTranslationKey(zoneId, fromResourceId);
-        String toKey = resourceKey(zoneId, toResourceId);
-        setResourceTranslation(fromKey, toKey);
-    }
+    private void setEntityTimestamps(final PolicyEvaluationRequestCacheKey key, final PolicyEvaluationResult result) {
+        // This will lower the overall hit rate for policy evaluation but is required to ensure that if
+        // the cached timestamp for any entity involved in this decision is aged out, the cached decision will be used
```
Comment (irinaepshteyn) [inline service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java]: Did you mean WILL NOT be used?

File: service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java [hunk 6]
```diff
@@ -67,6 +88,11 @@ public PolicyEvaluationResult get(final PolicyEvaluationRequestCacheKey key) {
         } catch (IOException e) {
             throw new IllegalStateException("Failed to read policy evaluation result as JSON.", e);
         }
+        values.remove(values.size() - 1);
```
Comment (irinaepshteyn) [inline service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java]: May be values.removeAll()?

File: service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java [hunk 7]
```diff
@@ -107,11 +135,10 @@ private void logCacheGetDebugMessages(final PolicyEvaluationRequestCacheKey key,
     @Override
     public void set(final PolicyEvaluationRequestCacheKey key, final PolicyEvaluationResult result) {
         try {
-            DateTime now = new DateTime();
-            result.setTimestamp(now.getMillis());
+            setEntityTimestamps(key, result);
+            result.setTimestamp(new DateTime().getMillis());
             String value = OBJECT_MAPPER.writeValueAsString(result);
```
Comment (irinaepshteyn) [inline service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java:220]: Add a unit test that caches the eval result with resolved URIs and get returns the valid cache since timestamps are recorded.

File: service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java (hunk 3 above)
Comment (irinaepshteyn) [reply service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java]: Add a unit test that verifies that get returns null if timestamps for any entity including resolved URIs are not present.

File: service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java [hunk 8]
```diff
@@ -150,139 +176,136 @@ public void reset(final PolicyEvaluationRequestCacheKey key) {
...
+        String key = getKey.apply(zoneId, entityId);
+        setIfNotExists(key, timestampValue());
+    }
+
+    private void logSetEntityTimestampsDebugMessage(final String timestamp, final String key, final String entityId,
+            final EntityType entityType) {
         if (LOGGER.isDebugEnabled()) {
-            LOGGER.debug(String.format("Setting timestamp for poliy set '%s'; key: '%s', value: '%s' ", policySetId,
-                    key, timestamp));
+            LOGGER.debug("Setting timestamp for {} '{}'; key: '{}', value: '{}'", entityType, entityId, key, timestamp);
```
Comment (irinaepshteyn) [inline service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java]: Since you are using lazy string construction we could rely on debug to check the debug level.  No need for extra isDebugEnabled call.

File: service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java [hunk 9]
```diff
@@ -150,139 +176,136 @@ public void reset(final PolicyEvaluationRequestCacheKey key) {
...
+        logSetEntityTimestampsDebugMessage(timestamp, key, entityId, entityType);
+        set(key, timestamp);
+    }
+
+    private void setEntityIfNotExists(final String zoneId, final String entityId,
+            final BiFunction<String, String, String> getKey) {
+        String key = getKey.apply(zoneId, entityId);
+        setIfNotExists(key, timestampValue());
+    }
+
```
Comment (irinaepshteyn) [inline service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java]: Is there a difference between setEntityIfNotExists and resetForEntity?
When eval result is being cached we could call resetForEntity to set the timestamp.  It will create a entity key if doesn't exist.

File: service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java [hunk 10]
```diff
@@ -150,139 +176,136 @@ public void reset(final PolicyEvaluationRequestCacheKey key) {
...
-        multisetForResources(zoneId, resourceIds.stream().collect(Collectors.toList()));
+        multisetForResources(zoneId, new ArrayList<>(resourceIds));
     }
 
     @Override
     public void resetForResources(final String zoneId, final List<ResourceEntity> resourceEntities) {
-        multisetForResources(zoneId, resourceEntities.stream().map(resource -> resource.getResourceIdentifier())
+        multisetForResources(zoneId, resourceEntities.stream().map(ResourceEntity::getResourceIdentifier)
                 .collect(Collectors.toList()));
     }
```
Comment (irinaepshteyn) [inline service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java]: We are collecting resource Ids into a list and then build a map of keys inside multisetForResources.  Could we combine it all in one call? Map streamed resourceEntities to resourceKey and collect into a map with the timestamp logging debug message for each... :-)

File: service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java [hunk 11]
```diff
@@ -150,139 +176,136 @@ public void reset(final PolicyEvaluationRequestCacheKey key) {
...
-        multisetForSubjects(zoneId, subjectIds.stream().collect(Collectors.toList()));
+        multisetForSubjects(zoneId, new ArrayList<>(subjectIds));
     }
 
     @Override
     public void resetForSubjects(final String zoneId, final List<SubjectEntity> subjectEntities) {
         multisetForSubjects(zoneId,
-                subjectEntities.stream().map(subject -> subject.getSubjectIdentifier()).collect(Collectors.toList()));
+                subjectEntities.stream().map(SubjectEntity::getSubjectIdentifier).collect(Collectors.toList()));
     }
```
Comment (irinaepshteyn) [inline service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java]: Same comment as for resetForResources above.  Could we combine resetForSubjects and multisetForSubjects in one long beautiful fluent java 8 call? :-)

File: service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java [hunk 12]
```diff
@@ -150,139 +176,136 @@ public void reset(final PolicyEvaluationRequestCacheKey key) {
...
 
-    boolean havePrivilegeServiceAttributesChanged(final List<String> values,
-            final DateTime policyEvalTimestampUTC) {
-        for (int i = 0; i < values.size() - 1; i++) {
-            if (null == values.get(i)) {
-                continue;
+    boolean havePrivilegeServiceAttributesChanged(final List<String> values, final DateTime policyEvalTimestampUTC) {
+        for (String value : values) {
+            if (null == value) {
+                return true;
```
Comment (irinaepshteyn) [inline service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java]: Values were already tested for nulls by isRequestCached before calling this method.

File: service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java [hunk 13]
```diff
@@ -150,139 +176,136 @@ public void reset(final PolicyEvaluationRequestCacheKey key) {
...
                 throw new IllegalStateException("Failed to read timestamp from JSON.", e);
             }
-
             if (invalidationTimestampUTC.isAfter(policyEvalTimestampUTC)) {
+                if (LOGGER.isDebugEnabled()) {
+                    LOGGER.debug(
+                            "Privilege service attributes have timestamp '%s' which is after "
+                                    + "policy evaluation timestamp '%s'",
+                            invalidationTimestampUTC, policyEvalTimestampUTC);
+                }
```
Comment (irinaepshteyn) [inline service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java]: Since you are using lazy string construction we could rely on debug to check the debug level. No need for extra isDebugEnabled call.  Will message formatting work with %s or should we use {}?

**Review by irinaepshteyn — COMMENTED** [review]
Review is still in progress.  These are the initial comments...

File: service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java (hunk 3 above)
Comment (irinaepshteyn) [reply service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java]: I take it back.  Resolved UIRs could not be part of the eval result key as they are not known at the time of the request.

File: service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java (hunk 9 above)
Comment (FrankGasparovic) [reply service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java]: setEntityIfNotExists will set the timestamp for the object if the object is not in the cache. resetForEntity will reset the timestamp regardless if the value if the value is in the cache

File: service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java (hunk 5 above)
Comment (FrankGasparovic) [reply service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java]: I do mean will, but I think the message is confusing. I rewrote it.

File: service/src/main/java/com/ge/predix/acs/service/policy/evaluation/PolicyEvaluationServiceImpl.java (hunk 1 above)
Comment (FrankGasparovic) [reply service/src/main/java/com/ge/predix/acs/service/policy/evaluation/PolicyEvaluationServiceImpl.java]: Fixed

File: service/src/main/java/com/ge/predix/acs/service/policy/evaluation/PolicyEvaluationServiceImpl.java (hunk 2 above)
Comment (FrankGasparovic) [reply service/src/main/java/com/ge/predix/acs/service/policy/evaluation/PolicyEvaluationServiceImpl.java:124]: Fixed

File: service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java (hunk 3 above)
Comment (FrankGasparovic) [reply service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java]: Fixed

File: service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java (hunk 7 above)
Comment (FrankGasparovic) [reply service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java:220]: testGetWithCacheHit covers this case

File: service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java (hunk 8 above)
Comment (FrankGasparovic) [reply service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java]: Removed

File: service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java [hunk 14]
```diff
@@ -67,6 +96,14 @@ public PolicyEvaluationResult get(final PolicyEvaluationRequestCacheKey key) {
...
             throw new IllegalStateException("Failed to read policy evaluation result as JSON.", e);
         }
+        // Remove the Policy Evaluation Result out of values because we deserialized it
+        values.remove(values.size() - 1);
+
+        String zone = key.getZoneId();
+        // Add the resolved resource URI's stored in the cachedResult to values so that we can check timestamps of
+        // all entities involved in the request.
+        values.addAll(multiGet(cachedResult.getResolvedResourceUris().stream()
+                .map(resourceUri -> resourceKey(zone, resourceUri)).collect(Collectors.toList())));
```
Comment (irinaepshteyn) [inline service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java]: One optimization we could do here is to compare resource from request with resolved resource.  If policy doesn't contain any template URIs then request resource and resolved resource are the same.  In that case we don't have to perform a multiGet again.  In that case when assembling keys for the first call to multiGet we need to include a resource from request again.

File: service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/PolicyEvaluationCache.java [hunk 15]
```diff
@@ -29,24 +29,26 @@
...
-
-    void setResourceTranslations(String zoneId, Set<String> fromResourceIds, String toResourceId);
-
     void reset();
 
     void reset(PolicyEvaluationRequestCacheKey key);
 
     void resetForPolicySet(String zoneId, String policySetId);
 
+    void setPolicySetIfNotExists(String zoneId, String policySetId);
```
Comment (irinaepshteyn) [inline service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/PolicyEvaluationCache.java]: setPolicySetIfNotExists, setResourceIfNotExists, setSubjectIfNotExists do not need to be part of cache interface.  These methods are implemented and used by AbstractPolicyEvaluationCache only.  However, setIfNotExists should be defined by interface and implemented by each cache implementation.

**Review by irinaepshteyn — COMMENTED** [review]
More comments...

File: service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/PolicyEvaluationCache.java (hunk 15 above)
Comment (FrankGasparovic) [reply service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/PolicyEvaluationCache.java]: I agree that setPolicySetIfNotExists, setResourceIfNotExists and setSubjectIfNotExists should not be part of the cache interface (because these are only used internally in the AbstractPolicyEvaluationCache). But why should setIfNotExists be apart of the interface if it is also only use internally?

File: service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/RedisPolicyEvaluationCache.java [hunk 16]
```diff
@@ -123,21 +97,7 @@ public void setCachedEvalTimeToLiveSeconds(final long cachedEvalTimeToLiveSecond
...
-            public List<Object> doInRedis(final RedisConnection connection) throws DataAccessException {
-                StringRedisConnection stringRedisConn = new DefaultStringRedisConnection(connection);
-                for (String fromKey : fromKeys) {
-                    stringRedisConn.sAdd(fromKey, toKey);
-                }
-                return null;
-            }
-        });
+    void setIfNotExists(final String key, final String value) {
+        this.redisTemplate.boundValueOps(key).setIfAbsent(value);
```
Comment (irinaepshteyn) [inline service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/RedisPolicyEvaluationCache.java:100]: Should we set an expiration timeout on the key similar to what's done in line #89?  The same comment applies to set function above.  Timeout is only set on the key for eval result and not for any other.

Comment (FrankGasparovic) [reply service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/RedisPolicyEvaluationCache.java:100]: Timeout should not be set for anything other then Policy Evaluation. This is because we would then be unnecessarily reevaluating requests when entities have no changed.

File: service/src/test/java/com/ge/predix/acs/policy/evaluation/cache/InMemoryPolicyEvaluationCacheTest.java [hunk 17]
```diff
@@ -73,9 +70,7 @@ public void testSetPolicyPolicySetChangedTimestamp() throws Exception {
         this.cache.set(key, value);
 
         Map<String, String> evalCache = (Map<String, String>) getInternalState(this.cache, "evalCache");
-        assertEquals(evalCache.size(), 0);
-        Map<String, String> timestampCache = (Map<String, String>) getInternalState(this.cache, "timestampCache");
-        assertEquals(timestampCache.size(), 1);
+        assertEquals(evalCache.size(), 1);
```
Comment (irinaepshteyn) [inline service/src/test/java/com/ge/predix/acs/policy/evaluation/cache/InMemoryPolicyEvaluationCacheTest.java:73]: May be rename the test to testSetPolicySetChangedTimestamp?

File: service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java (hunk 14 above)
Comment (FrankGasparovic) [reply service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java]: Added

File: service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java (hunk 13 above)
Comment (FrankGasparovic) [reply service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java]: Fixed

File: service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java (hunk 11 above)
Comment (FrankGasparovic) [reply service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java]: Fixed

File: service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java (hunk 10 above)
Comment (FrankGasparovic) [reply service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java]: Fixed

File: service/src/test/java/com/ge/predix/acs/policy/evaluation/cache/InMemoryPolicyEvaluationCacheTest.java (hunk 17 above)
Comment (irinaepshteyn) [reply service/src/test/java/com/ge/predix/acs/policy/evaluation/cache/InMemoryPolicyEvaluationCacheTest.java:73]: Current implementation of the in-memory cache stores eval result and timestamps in the same map, not in different maps as before.  There is no need to have different tests for setting eval result and setting timestamps.

File: service/src/test/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCacheTest.java [hunk 18]
```diff
@@ -88,9 +91,13 @@ public void testGetWithCacheMiss() {
     public void testGetWithCacheHit() {
```
Comment (irinaepshteyn) [inline service/src/test/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCacheTest.java:108]: Add a test to assert that get returns cached result if eval result is cached but not timestamps for entities.

File: service/src/test/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCacheTest.java [hunk 19]
```diff
@@ -119,14 +129,18 @@ public void testGetWithPolicyInvalidation() throws Exception {
...
 
     @Test
     public void testGetWithResetForMultiplePolicySets() throws Exception {
 
         PolicyEvaluationRequestV1 request = new PolicyEvaluationRequestV1();
+        this.cache.resetForSubject(ZONE_NAME, AGENT_MULDER);
+        this.cache.resetForResource(ZONE_NAME, XFILES_ID);
+        this.cache.resetForPolicySet(ZONE_NAME, POLICY_ONE.getName());
+        this.cache.resetForPolicySet(ZONE_NAME, POLICY_TWO.getName());
         request.setAction(ACTION_GET);
```
Comment (irinaepshteyn) [inline service/src/test/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCacheTest.java:161]: Do we really need to call resets for each entity i every test?  Get will return cached result if no timestamps.

**Review by irinaepshteyn — CHANGES_REQUESTED** [review]
Done reviewing.

File: service/src/test/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCacheTest.java (hunk 19 above)
Comment (FrankGasparovic) [reply service/src/test/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCacheTest.java:161]: We do, get will return null if no timestamps.

File: service/src/test/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCacheTest.java (hunk 18 above)
Comment (FrankGasparovic) [reply service/src/test/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCacheTest.java:108]: If eval is cached but timestamps for entities are not, then get should return null

File: service/src/test/java/com/ge/predix/acs/policy/evaluation/cache/InMemoryPolicyEvaluationCacheTest.java (hunk 17 above)
Comment (FrankGasparovic) [reply service/src/test/java/com/ge/predix/acs/policy/evaluation/cache/InMemoryPolicyEvaluationCacheTest.java:73]: I do feel like there is value in this test because of the logic to check the key format.

**Comment by FrankGasparovic** [comment]
Siva ran performance tests and he saw a 3ms increase in Policy Evaluation calls

**Comment by sanjeevchopra** [comment]
@anubhavi25 @FrankGasparovic - pls review my 2nd commit

File: service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java [hunk 20]
```diff
@@ -80,59 +79,122 @@ public String toString() {
...
+        PolicyEvaluationResult cachedEvalResult = toPolicyEvaluationResult(cachedEvalResultString);
+        
+        List<String> invalidationTimeStamps = new ArrayList<>();
+        invalidationTimeStamps.add(cachedEntries.getSubjectLastModified());
+        invalidationTimeStamps.addAll(cachedEntries.getPolicySetsLastModified());
+
+        Set<String> cachedResolvedResourceUris = cachedEvalResult.getResolvedResourceUris();
+        //is requested resource id same as resolved resource uri ?
+        if (cachedResolvedResourceUris.size() == 1
+                && cachedResolvedResourceUris.toArray()[0].equals(key.getResourceId())) {
```
Comment (FrankGasparovic) [inline service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java]: Use cachedResolvedResourceUris.iterator().next() vs toArray() so that we don't have to copy contents to array.

**Review by FrankGasparovic — COMMENTED** [review]
Looks good.

File: service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java [hunk 21]
```diff
@@ -80,59 +79,122 @@ public String toString() {
...
-    }
+        String getDecisionKey() {
+            return decisionKey;
+        }
+
+        String getSubjectLastModified() {
+            return entryValues.get(lastValueIndex - 2);
+        }
+
+        String getRequestedResourceLastModified() {
```
Comment (anubhavi25) [inline service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java:189]: Changing the name to just `getResourceLastModified` should suffice (and is also consistent with the naming convention used for the other `get*LastModified` methods).

**Review by anubhavi25 — COMMENTED** [review]
Looks good for the most part 👍

**Comment by sanjeevchopra** [comment]
Performance tests show no regression. A second test was run which only writes resource and subject keys (which have no TTL) to confirm that eviction is working as expected.
